pkg/daemon/on_disk_validation.go

```diff
@@ -0,0 +1,250 @@
+	}
+}
+
+// Checks that the ondisk state for a systemd dropin matches the expected state.
+func checkV3Dropin(systemdPath string, unit ign3types.Unit, dropin ign3types.Dropin) error {
+	path := getIgn3SystemdDropinPath(systemdPath, unit, dropin)
+
+	var content string
+	if dropin.Contents == nil {
+		content = ""
+	} else {
+		content = *dropin.Contents
+	}
+
+	// As of 4.7 we now remove any empty defined dropins, check for that first
+	if _, err := os.Stat(path); content == "" && err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		}
+		return err
+	}
+
+	// To maintain backwards compatibility, we allow existing zero length files to exist.
+	// Thus we are also ok if the dropin exists but has no content
+	return checkFileContentsAndMode(path, []byte(content), defaultFilePermissions)
+}
+
+// checkUnits validates the contents of all the units in the
+// target config and returns nil if they match.
+func checkV3Units(units []ign3types.Unit, systemdPath string) error {
+	for _, unit := range units {
+		for _, dropin := range unit.Dropins {
+			if err := checkV3Dropin(systemdPath, unit, dropin); err != nil {
+				return err
+			}
+		}
+
+		if unit.Contents == nil || *unit.Contents == "" {
+			continue
+		}
+
+		path := getIgn3SystemdUnitPath(systemdPath, unit)
+		if unit.Mask != nil && *unit.Mask {
+			link, err := filepath.EvalSymlinks(path)
+			if err != nil {
+				return errors.Wrapf(err, "state validation: error while evaluation symlink for path %q", path)
+			}
+			if strings.Compare(pathDevNull, link) != 0 {
+				return errors.Errorf("state validation: invalid unit masked setting. path: %q; expected: %v; received: %v", path, pathDevNull, link)
+			}
+		}
+		if err := checkFileContentsAndMode(path, []byte(*unit.Contents), defaultFilePermissions); err != nil {
+			return err
+		}
+
+	}
+	return nil
+}
+
+// checkUnits validates the contents of all the units in the
+// target config and returns nil if they match.
+func checkV2Units(units []ign2types.Unit, systemdPath string) error {
+	for _, unit := range units {
+		for _, dropin := range unit.Dropins {
+			path := getIgn2SystemdDropinPath(systemdPath, unit, dropin)
+			if err := checkFileContentsAndMode(path, []byte(dropin.Contents), defaultFilePermissions); err != nil {
+				return err
+			}
+		}
+
+		if unit.Contents == "" {
+			continue
+		}
+
+		path := getIgn2SystemdUnitPath(systemdPath, unit)
+		if unit.Mask {
+			link, err := filepath.EvalSymlinks(path)
+			if err != nil {
+				return errors.Wrapf(err, "state validation: error while evaluation symlink for path %q", path)
+			}
+			if strings.Compare(pathDevNull, link) != 0 {
+				return errors.Errorf("state validation: invalid unit masked setting. path: %q; expected: %v; received: %v", path, pathDevNull, link)
+			}
+		}
+		if err := checkFileContentsAndMode(path, []byte(unit.Contents), defaultFilePermissions); err != nil {
+			return err
+		}
+
+	}
+	return nil
+}
+
+// checkFiles validates the contents of  all the files in the
```

Comment:
instead of `checkFiles` , it can be `check3Files` to match comment with function.